### PR TITLE
test data for fixed and dynamic huffman trees

### PR DESCRIPTION
**Pull request by garymm**
test data for fixed and dynamic huffman trees

and a script to generate it

I guess the nice way to do this would be to bring in rules_python and
generate the data in a genrule, or pull in zlib and have our test
generate the compressed data at run-time. We can add that later if
we find keeping the compressed and decompressed data in sync is annoying
enough.

Change-Id: Iaba4043956a2eb6fd0f51aef40414ae948a7311e

### CODE REVIEW

**Comment by oliverlee** [comment]
Can you add the genrule and a diff test so the generated files are always in sync? 

https://github.com/bazelbuild/bazel-skylib/blob/main/docs/diff_test_doc.md

I think it's fine if we aren't using rules_python and a hermetic interpreter + packages.

**Comment by oliverlee** [comment]
I think it should be fairly straightforward and can probably do it later tonight if you're not familiar with that.

File: tools/compressed_file.bzl [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+Compresses a DEFLATE compressed file with a fixed or dynamic Huffman tree.
+"""
+
+def compressed_file(
+        name,
+        src,
+        out = None,
+        strategy = "fixed"):
+    """
+    Compresses a file with the DEFLATE algorithms and a fixed or dynamic Huffman tree.
```
Comment (oliverlee) [inline tools/compressed_file.bzl:11]:
```suggestion
    Compresses a file with the DEFLATE algorithm and a fixed or dynamic Huffman tree.
```

**Review by oliverlee — COMMENTED** [review]
@garymm not sure I interpreted the compression stuff right, feel free to change whatever.

It's good to know that I generated the same files you did on your machine.

**Review by oliverlee — APPROVED** [review]

**Review by garymm — COMMENTED** [review]
I didn't like the diff tests, so I just removed the checked-in files.
I did have some trouble reading the files into std::byte though. PTAL

File: src/test/decompress_test.cpp [hunk 2]
```diff
@@ -11,7 +15,32 @@ constexpr auto byte_vector(Ts... values)
...
+  std::ifstream file{abs_path, std::ios::binary};
+  ::boost::ut::expect(::boost::ut::fatal(file.is_open()))
+      << "failed to open " << path;
+  std::vector<char> chars(
+      (std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
+
+  const std::vector<char> char_vector(
+      (std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
+
+  // The standard library really doesn't want me to read bytes froma file, so
```
[the remark this answers is not in the document]
Comment (oliverlee) [reply src/test/decompress_test.cpp:37]: I'll try and take a look later today

**Comment by oliverlee** [comment]
wrote some commits, should probably figure out how to split into different PRs and merge

File: src/test/BUILD.bazel [hunk 3]
```diff
@@ -21,14 +21,18 @@ cc_test(
     name = "decompress_test",
     timeout = "short",
     srcs = ["decompress_test.cpp"],
+    args = [
+        "$(rootpath :starfleet.html.fixed)",
```
Comment (garymm) [inline src/test/BUILD.bazel:25]: nice!

**Review by garymm — COMMENTED** [review]
All looks great except for the bytes file I/O question which is in its own PR.

**Comment by garymm** [comment]
I accidentally dropped some of your commits in the process of local pulling and pushing, but I'd like to get this in to make progress. I think this is decent enough to merge.

**Pull request merged**